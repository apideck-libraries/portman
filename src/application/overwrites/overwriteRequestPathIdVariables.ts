import pluralize from 'pluralize'
import { PostmanMappedOperation } from 'postman'
import { OverwritePathIdVariableConfig } from 'types'

/**
 * Overwrite Postman request path variables with values defined by the portman testsuite
 * @param overwriteValues
 * @param pmOperation
 */
export const overwriteRequestPathIdVariables = (
  overwriteValues: OverwritePathIdVariableConfig[],
  pmOperation: PostmanMappedOperation
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
): PostmanMappedOperation => {
  // Early exit if overwrite values are not defined
  if (!(overwriteValues instanceof Array)) return pmOperation

  // Early exit if request path variables are not defined
  if (!pmOperation.item?.request) return pmOperation

  // Early exit if request name/resource is not defined
  if (!pmOperation.item?.name) return pmOperation

  const pathParams = pmOperation.item.request.url?.path
    ? [...pmOperation.item.request.url?.path]
    : []

  const idIndex = pathParams.indexOf(':id')

  // Early exit if path variables do not include :id
  if (idIndex < 0) return pmOperation

  // Assume the resource is defined restfully :-/
  const name = pathParams[idIndex - 1]
  const singular = pluralize.singular(name.toLowerCase())
  const variableName = `{{${singular}Id}}`

  pmOperation.item.request.url.variables.map(item => {
    if (item.key === 'id') {
      item.value = variableName
    }
  })

  return pmOperation
}
