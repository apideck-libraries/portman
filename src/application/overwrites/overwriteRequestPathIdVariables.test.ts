import {
  getPostmanMappedCreateOperation,
  getPostmanMappedOperation
} from 'testUtils/getPostmanMappedOperation'
import { overwriteRequestPathIdVariables } from './overwriteRequestPathIdVariables'

describe('overwriteRequestPathIdVariables', () => {
  it('should overwrite the request path id variable with resource name', async () => {
    const overwriteValues = [
      {
        enabled: true
      }
    ]
    const pmOperation = await getPostmanMappedOperation()
    const result = overwriteRequestPathIdVariables(overwriteValues, pmOperation)
    expect(result.item.request.url.variables).toMatchSnapshot()
  })

  it('should return unaltered if id not found', async () => {
    const overwriteValues = [
      {
        enabled: true
      }
    ]

    const pmOperation = await getPostmanMappedCreateOperation()
    const result = overwriteRequestPathIdVariables(overwriteValues, pmOperation)

    expect(result.item.request.url.variables).toEqual(pmOperation.item.request.url.variables)
  })
})
