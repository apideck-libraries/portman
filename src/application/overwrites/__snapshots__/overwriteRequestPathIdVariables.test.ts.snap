// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overwriteRequestPathIdVariables should overwrite the request path id variable with resource name 1`] = `
Array [
  Object {
    "description": Object {
      "content": "(Required) ID of the record you are acting upon.",
      "type": "text/plain",
    },
    "disabled": false,
    "key": "id",
    "type": "any",
    "value": "{{companyId}}",
  },
]
`;
